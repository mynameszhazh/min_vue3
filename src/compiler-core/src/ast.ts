export const enum NodeTypes {
  INTERPOLATION,
  TEXT,
  SIMPLE_EXPRESSION,
  ELEMENT,
}
