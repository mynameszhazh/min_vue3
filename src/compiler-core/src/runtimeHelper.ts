export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREAT_ELEMENT_VNODE = Symbol("createElementVNode");

export const helperMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [ CREAT_ELEMENT_VNODE ]: 'createElementVNode'
};
