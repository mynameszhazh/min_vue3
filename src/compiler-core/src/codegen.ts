import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREAT_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelper";

export function genarator(ast) {
  let context: any = createCodegenContext();
  let { push } = context;

  genFunctionPreamble(context, ast);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${functionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code,
  };
}
function genFunctionPreamble(context: any, ast: any) {
  const VueBinging = "Vue";
  let { push } = context;
  let aliasHelpers = (s) => `${helperMapName[s]}: ${helperMapName[s]}`;
  if (ast.helpers.length > 0) {
    push(
      `const { ${ast.helpers.map(aliasHelpers).join(", ")} } = ${VueBinging}`
    );
  }
  push("\n");
  push("return ");
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterprolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
    default:
      break;
  }
}
function genElement(node, context) {
  const { push, helper } = context;
  const { tag, props, children } = node;
  push(`${helper(CREAT_ELEMENT_VNODE)}('${tag}', null, `);
  // push(`${helper(CREAT_ELEMENT_VNODE)}(`);
  // genNodeList(genNullableArgs([tag, props, children]), context);

  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    genNode(child, context)
  }

  push(")");
}

function genNodeList(nodes: any, context: any) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];

    if (isString(node)) {
      push(`${node}`);
    } else {
      genNode(node, context);
    }
    // node 和 node 之间需要加上 逗号(,)
    // 但是最后一个不需要 "div", [props], [children]
    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullableArgs(args) {
  // 把末尾为null 的都删除掉
  // vue3源码中，后面可能会包含 patchFlag、dynamicProps 等编译优化的信息
  // 而这些信息有可能是不存在的，所以在这边的时候需要删除掉
  let i = args.length;
  // 这里 i-- 用的还是特别的巧妙的
  // 当为0 的时候自然就退出循环了
  while (i--) {
    if (args[i] != null) break;
  }

  // 把为 falsy 的值都替换成 "null"
  return args.slice(0, i + 1).map((arg) => arg || "null");
}

function createCodegenContext(): { push: any } {
  let context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}
function genText(node: any, context: any) {
  let { push } = context;
  push(`'${node.content}'`);
}

function genInterprolation(node: any, context: any) {
  let { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genExpression(node: any, context: any) {
  let { push } = context;
  push(`${node.content}`);
}
