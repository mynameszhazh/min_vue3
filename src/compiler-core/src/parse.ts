import { NodeTypes } from "./ast";

const enum TagType {
  START,
  END,
}

export function baseParse(content) {
  const context = createParserContext(content);
  return createRoot(parseChildren(context));
}

function createParserContext(content: any) {
  return {
    source: content,
  };
}

function createRoot(content) {
  return {
    children: content,
  };
}

function parseChildren(context: any): any {
  const nodes: any = [];
  let node;
  const s = context.source;
  if (s.startsWith("{{")) {
    // 解析插值
    node = parseInterplation(context);
  } else if (s.startsWith("<")) {
    if (/[a-z]/i.test(s[1])) {
      console.log("parse elemetn");
      node = parseElement(context);
    }
  }

  nodes.push(node);
  return nodes;
}

function parseInterplation(context): any {
  // context  = {{message}}
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  advanceBy(context, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;
  const rawContent = context.source.slice(0, rawContentLength);
  const content = rawContent.trim();

  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function parseElement(context: any): any {
  // 获取 tag
  let element = parseTag(context, TagType.START);
  parseTag(context, TagType.END);
  return element;
}

function parseTag(context: any, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  advanceBy(context, match[0].length);
  advanceBy(context, 1);

  if(type === TagType.END) return

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function advanceBy(context: any, length: number): any {
  context.source = context.source.slice(length);
}
