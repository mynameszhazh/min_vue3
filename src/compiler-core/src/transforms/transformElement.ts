import { createVnodeCall, NodeTypes } from "../ast";

export function transformElment(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      // 中间处理层

      // tag
      const vnodeTag = `'${node.tag}'`;

      // props
      let vnodeProps;

      // children
      const children = node.children;
      let vnodeChildren = children[0];

      node.codegenNode = createVnodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
