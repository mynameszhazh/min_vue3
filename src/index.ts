export * from "./runtime-dom/index";

import { baseCompaile } from "./compiler-core/src";
import * as runtimeDom from "./runtime-dom/index";
import { registerRuntimeCompiler } from "./runtime-dom/index";
function compaileToFunction(template) {
  const { code } = baseCompaile(template);

  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compaileToFunction);
