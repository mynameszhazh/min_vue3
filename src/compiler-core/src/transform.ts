import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelper";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);

  createCodegenNode(root);

  root.helpers = [...context.helpers.keys()];
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
    nodeTransforms: options.nodeTransforms || [],
  };
  return context;
}
function traverseNode(node: any, context: any) {
  // 执行一些指定的执行操作
  let nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    let transform = nodeTransforms[i];
    // 这里执行一些外部需要执行的插件
    transform(node, context);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      // 引入一些导入的内容
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
      traverChildNode(node, context);
      break;
    case NodeTypes.ELEMENT:
      traverChildNode(node, context);
      break;
    default:
      break;
  }
}

function traverChildNode(node: any, context) {
  let children = node.children;
  for (let i = 0; i < children.length; i++) {
    const childNode = children[i];
    traverseNode(childNode, context);
  }
}
function createCodegenNode(root) {
  root.codegenNode = root.children[0];
}
