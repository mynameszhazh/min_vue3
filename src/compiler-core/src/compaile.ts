import { genarator } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElment } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

export function baseCompaile(template) {
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElment, transformText],
  });
  return genarator(ast);
}
