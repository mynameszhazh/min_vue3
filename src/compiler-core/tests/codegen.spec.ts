import { genarator } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElment } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

describe("code gen", () => {
  it("happy path", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = genarator(ast);

    // 快照
    // 1. 抓bug
    // 2. 有意
    expect(code).toMatchSnapshot();
  });

  it("innser codeGen", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      // 用到插值 才使用transformExpression
      nodeTransforms: [transformExpression],
    });
    const { code } = genarator(ast);

    expect(code).toMatchSnapshot();
  });
  it.only("element", () => {
    const ast = baseParse("<div>hi, {{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElment, transformText],
    });
    const { code } = genarator(ast);

    expect(code).toMatchSnapshot();
  });
});
