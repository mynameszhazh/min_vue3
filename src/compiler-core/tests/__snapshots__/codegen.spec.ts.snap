// Vitest Snapshot v1

exports[`code gen > element 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`code gen element 1`] = `
"const { toDisplayString: toDisplayString, creatElementVnode: creatElementVnode } = Vue
return function render(_ctx, _cache){return _creatElementVnode('div', null, 'hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`code gen happy path 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;

exports[`code gen innser codeGen 1`] = `
"const { toDisplayString: toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;
